ML/ml_training_xor_application.py

```diff
@@ -0,0 +1,599 @@
+        self.vars_to_draw = config_train["plots"]["extra_columns"] + self.training_vars
+
+        self.raw_output = config_train["training"]["raw_output"]
+        self.roc_auc_approach = config_train["training"]["roc_auc_approach"]
+        self.roc_auc_average = config_train["training"]["roc_auc_average"]
+        self.score_metric = "roc_auc"
+        self.hyper_pars = config_train["training"]["hyper_pars"]
+        self.hyper_pars_opt = config_train["training"]["hyper_pars_opt"]
+
+        # output
+        self.outdir = config_train["output"]["dir"]
+        self.extension = config_train["plots"]["extension"]
+        self.log_file = config_train["output"]["log_file"]
+
+    def __check_input_consistency(self):
+        """
+        Helper method to check self consistency of inputs
+        """
+
+        # class balance
+        if self.share not in ("equal", "all_signal"):
+            print(f"\033[91mERROR: class_balance option {self.share} not implemented\033[0m")
+            sys.exit()
+        if self.share == "all_signal" and len(self.bkg_factor) != len(self.pt_bins):
+            print("\033[91mERROR: bkg_factor must be defined for each pT bin!\033[0m")
+            sys.exit()
+        # training
+        if self.training_vars is None:
+            print("\033[91mERROR: training columns must be defined!\033[0m")
+            sys.exit()
+        # hyper-parameters options
+        if not isinstance(self.hyper_pars, list):
+            print("\033[91mERROR: hyper-parameters must be defined or be a list containing an empty dict!\033[0m")
+            sys.exit()
+        if not isinstance(self.hyper_pars[0], dict):
+            print("\033[91mERROR: hyper-parameters must be a list of dict!\033[0m")
+            sys.exit()
+        if len(self.hyper_pars) != len(self.pt_bins):
+            print("\033[91mERROR: hyper-parameters definition does not match pT binning!\033[0m")
+            sys.exit()
+        if not isinstance(self.hyper_pars_opt["hyper_par_ranges"], dict):
+            print("\033[91mERROR: hyper_pars_opt_config must be defined!\033[0m")
+            sys.exit()
+
+    def __get_sliced_dfs(self):
+        """
+        Helper method to get pT-sliced dataframes for each class
+
+        Returns
+        -----------------
+        - hdl_bkg: pandas dataframe containing only background candidates
+        - hdl_sig: pandas dataframe containing only signal candidates
+        """
+
+        print("Loading and preparing data files: ...", end="\r")
+
+        hdl = TreeHandler(file_name=self.infile_name, tree_name=self.tree_name, folder_name=self.folder_name)
+
+        hdl_bkg = hdl.get_subset(f"{self.tag} == 0 and {self.filt_bkg_mass}")
+        hdl_sig = hdl.get_subset(f"{self.tag} == 1")
+
+        hdl_bkg.slice_data_frame(self.name_pt_var, self.pt_bins, True)
+        hdl_sig.slice_data_frame(self.name_pt_var, self.pt_bins, True)
+
+        print("Loading and preparing data files: Done!")
+        return hdl_bkg, hdl_sig
```

Comment:
Thank you @AlexBigO for the nice code! I was wondering if it would be easy to add the possibility of training on MC signal examples and data bkg examples. Maybe one can add a data_file_name on the config, which can be left as null, and in that case the MC bkg is used for the training? But it's up to you if you wish to add it in a second iteration :)

Reply:
Ah but yes, of course! Thanks for spotting this, I missed this point when I tried to mix the preparation and training macros :)
I am adding a new configurable to be able to select bkg from real data or MC